src/flag_gems/utils/tensor_wrapper.py

```diff
@@ -0,0 +1,79 @@
+import math
+
+import torch
+
+
+class TypedPtr:
+    """This is a minimal requirement for a type to be treated as a tensor in triton jit
+    function. Basically it is ia typed pointer, withou knowning the device, size, shape,
+    strides, etc.
+    """
+
+    def __init__(self, ptr: int, dtype: torch.dtype):
+        self.ptr = ptr
+        self.dtype = dtype
+
+    def data_ptr(self) -> int:
+        return self.ptr
+
+    @classmethod
+    def from_tensor(cls, tensor: torch.Tensor, offset: int = 0):
+        return cls(tensor.data_ptr() + tensor.element_size() * offset, tensor.dtype)
+
+    @classmethod
+    def reinterpret_tensor(cls, tensor: torch.Tensor, dtype: torch.dtype, offset=0):
+        return cls(tensor.data_ptr() + dtype.itemsize * offset, dtype)
+
+
+class StridedBuffer:
+    """A drop in replacement of torch.Tensor that can be used in wrapper generated by
+    PointwiseDynamicFunction. It allows us to use a different shape, stride, data
+    pointer as the base tensor.
+
+    It is a kind of reinterpretation of the base tensor. We make this class since we
+    cannot get a Tensor view with negative strides via torch APIs, while we need this
+    to implement flip op.
+
+    Although generated code can accept torch.Tensor & StridedBuffer, but StridedBuffer
+    may not have all the methods as torch.Tensors do. We add some attributes & methods
+    with the same name as torch.Tensor, which are used in the generated code. But we
+    may not cover all the methods, add one if what you need is missing here.
+
+    And can also be used in triton kernels since it also has dtype & data_ptr().
+    """
+
+    def __init__(
+        self, base: torch.Tensor, shape=None, strides=None, dtype=None, offset=0
+    ):
+        self._base = base
+        self.dtype = dtype or base.dtype
+        if offset == 0:
+            self._data_ptr = self._base.data_ptr()
+        else:
+            offset = self.dtype.itemsize * offset
```

Comment:
The offset param is only used in flip? we have shape and stride info to compute offset

Reply:
For now this is only used in flip, since we cannot construct a tensor view with negative strides via torch APIs. In many other cases, we can use torch's view ops to construct a view.